Fix documentation source links

ValerianRey opened the pull request:
## What does this PR do?

Fixes #1394

My solution is basically a copy & adaptation of the most recent version of numpy's solution (https://github.com/numpy/numpy/blob/master/doc/source/conf.py#L424). I think the previous version was also adapted from numpy, according to the comment in the code, but it was from an older numpy version and it had some problems.

Note that the code:
```python
# strip decorators, which would resolve to the source of the decorator
obj = unwrap(obj)
```
Fixes another issue that was not reported as far as I know (the inspect module failed to return the right file name when the function was wrapped in a decorator). An example of this is: [Metric.sync_context](https://torchmetrics.readthedocs.io/en/stable/references/metric.html?highlight=sync_context) which produces an even more bugged link than other source code links.

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**? No test necessary

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Yes!

ValerianRey commented:
From the output of [Job: make-docs](https://github.com/Lightning-AI/metrics/actions/runs/3919066817/jobs/6699842271), the generated links seem to be good when doc is generated by the CI/CD pipeline.
I have no idea how to test if the docs generated by read the docs will be good too.
Any idea how to test that @Borda?
I'm scared it won't work because originally there was some extra logic in this scenario:
```python
 if any(s in fname for s in ("readthedocs", "rtfd", "checkouts")):
     # /home/docs/checkouts/readthedocs.org/user_builds/pytorch_lightning/checkouts/
     #  devel/pytorch_lightning/utilities/cls_experiment.py#L26-L176
     path_top = os.path.abspath(os.path.join("..", "..", "..", ".."))
     fname = os.path.relpath(fname, start=path_top)
```
I replaced everything with Numpy's solution without really understanding this part.

I also have to remove the prints that I added for testing.

ValerianRey commented:
Update : I just realized that there was a readthedocs job that is done on this pull request. After a quick vérification to the generated docs, it seems to work (links seem to be good, at least for master). I'll make further verifications tomorrow, remove the prints, and mark the pull request as not draft anymore.

ValerianRey commented:
I tested with `torchmetrics.__version__ = "0.11.0"` and it works too (with links pointing to the v0.11.0 tag instead of master).
I also removed the prints.
So I think this is ready.

Borda reviewed (commented):
That look good, can we just wrap these code blocks to small functions so that from the name it is easier to follow what it is doing...

ValerianRey commented on docs/source/conf.py:
```diff
@@ -315,39 +315,64 @@ def package_list_from_file(file):
+    else:
+        line_str = ""
+
+    if "dev" in torchmetrics.__version__:
```
Yes. Should we just replace with the following code? 
```python
if any(s in torchmetrics.__version__ for s in ("dev", "rc")):
    version_str = "master"
else:
    version_str = f"v{torchmetrics.__version__}"
```
The documentation generated with `rc` and `dev` is always the documentation of the `latest` version on ReadTheDocs right? In this case I think it would be fine if the docs for `rc` and `dev` versions link to the code from `master`.

ValerianRey replied:
Should I push that change? @Borda

Borda replied:
could you, or send it in another pr?

ValerianRey replied:
Done. I tested locally with `torch.__version__ = "0.11.0"`, `torch.__version__ = "0.12.0rc"` and `torch.__version__ = "0.12.0dev"` and it seems to work in all cases.

ValerianRey commented:
I updated the code to split the different parts into functions. In the process of doing so, I realized that some conditions and try/except blocks were never used, and since it wasn't very clear why they were there I just removed them.
I hope the code is easier to read now!

stancld commented on docs/source/conf.py:
```diff
@@ -315,39 +315,58 @@ def package_list_from_file(file):
-            fname = "master/" + os.path.relpath(fname, start=os.path.abspath(".."))
-        source, lineno = inspect.getsourcelines(obj)
-        return fname, lineno, lineno + len(source) - 1
+    # try to find the file and line number, based on code from numpy:
+    # https://github.com/numpy/numpy/blob/master/doc/source/conf.py#L424
```
Shouldn't we be here more explicit w/ copyright as it's copied/adapted from numpy, or is this enough? I'm often a bit unsure with this. Otherwise 👌
